## Question 1
a. Floyd’s algorithm.
b. Depth First Search Algorithm.
c. augmenting path algorithm of Ford and Fulkerson

## Question 2
a. O(n).
b. O(n^2).
c. O(m+nlogm)
d. O((m + n)m/n)

## Question 3
a. 1

b. dL or $ln(n)L/ln(L+1)$

c. $k(1+k-1)+(k-1)^2+...+(k-1)^{d-1}）=n \\
     \frac{(k-1)^d-1}{k-2}k=n \\
     So: d= ln(\frac{n(k-2)}{k}+1)/ln(k-1)$
     The diameter:$2d=2ln(\frac{n(k-2)}{k}+1)/ln(k-1)$
     
d. All there network are.

## Question 4



In [21]:
import numpy as np
a=[16,17,10,26,13,14,28,45,10,13,12,10,136,16,25,36,12,14,22,10]


In [22]:
sum=0
N=20
for i in a:
    sum=sum+np.log(i/(10-1/2))
alpha=1+N*(sum)**(-1)

In [23]:
alpha

2.5216325745167225

In [24]:
sigma=(alpha-1)/np.sqrt(N)
sigma

0.3402473873397406

The $\alpha$ is 2.52; The error is 0.34.

## Question 5
a. The right network is scale free because its log-log plot is nearly straight so that it follows the power law.  
b. When the degree is 1000, the frequency is 0.0001. So the $\alpha=-ln(0.0001)/ln(1000)+1=2.333$

In [25]:
-np.log(0.0001)/np.log(1000)+1

2.333333333333333

In [26]:
np.exp(np.log(0.5)/(2.333-2)*(2.333-1))

0.062370040112628575

c. according to $W=P^{(\alpha-2)(\alpha-1)}$. ~6% of highest-degree nodes have a half of all the edge-ends.